motionplan/tpSpaceRRT.go

```diff
@@ -0,0 +1,435 @@
+//go:build !windows
+
+package motionplan
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/edaniels/golog"
+	"github.com/golang/geo/r3"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/motionplan/tpspace"
+	"go.viam.com/rdk/referenceframe"
+	"go.viam.com/rdk/spatialmath"
+)
+
+var (
+	defaultGoalCheck = 5
+	defaultAutoBB    = 1. // Automatic bounding box on driveable area as a multiple of start-goal distance
+
+	// whether to add intermediate waypoints.
+	defaultAddInt = false
+	// Add a subnode every this many mm along a valid trajectory. Large values run faster, small gives better paths
+	// Meaningless if the above is false.
+	defaultAddNodeEvery = 50.
+	defaultSearchRadius = 50.
+)
+
+// newCBiRRTMotionPlannerWithSeed creates a cBiRRTMotionPlanner object with a user specified random seed.
+func newTPSpaceMotionPlanner(
+	frame referenceframe.Frame,
+	seed *rand.Rand,
+	logger golog.Logger,
+	opt *plannerOptions,
+) (motionPlanner, error) {
+	if opt == nil {
+		return nil, errNoPlannerOptions
+	}
+	mp, err := newPlanner(frame, seed, logger, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tpspaceRRTMotionPlanner{
+		planner:   mp,
+		goalCheck: defaultGoalCheck,
+		autoBB:    defaultAutoBB,
+
+		addNodeEvery: defaultAddNodeEvery,
+
+		searchRadius:    defaultSearchRadius,
+		addIntermediate: defaultAddInt,
+	}, nil
+}
+
+// tpspaceRRTMotionPlanner.
+type tpspaceRRTMotionPlanner struct {
+	*planner
+	goalCheck int // Check if goal is reachable every this many iters
+
+	// size of bounding box within which to randomly sample points
+	// TODO: base this on frame limits?
+	autoBB       float64
+	addNodeEvery float64
+	searchRadius float64
+
+	addIntermediate bool
+}
+
+type candidate struct {
+	dist      float64
+	treeNode  *configurationNode
+	newNode   *configurationNode
+	finalGoal bool
+	err       error
+}
+
+// TODO: seed is not immediately useful for TP-space.
```

Comment:
Inputs are TP coordinates, so refer to a point on a *single* trajectory. TP-space motion plan problems must be posed in terms of relative positioning of the robot, so the query frame by definition starts at `NewZeroPose()`

Possibly `seed`, and most other places where we have `[]referenceframe.Input`, should in fact be replaced by `node`, which possibly should in fact be `Node`

Reply:
This would be a great addition.  Though if we are going to make it public we should probably change the name from Node to something else as that is very vague and wont make a lot of sense outside the context of this package

Reply:
We also should look at both `node` and `State` and see what we want to do with those.

Reply:
Probably not a thing to do as part of *this* pr